# PANdeMaiz — Entrenamiento 1D-CNN para Detección Sísmica

**Propósito:** Entrenar una red neuronal liviana para clasificar ventanas de 4 s
como **Ruido (0)** o **Sismo (1)**, a partir de los espectrogramas generados en
`seismic_dataset_builder_v3.ipynb`.

| Parámetro | Valor |
|-----------|-------|
| Input shape | `(65, 11, 3)` float32 |
| Clases | 0 = Ruido (3 284 muestras) · 1 = Sismo (2 004 muestras) |
| Arquitectura | 1D-CNN (reshape frec→feature, Conv1D × 2) |
| Target tamaño | < 300 KB INT8 TFLite |
| Target recall sismo | ≥ 85 % en test set |
| PGA | No incluido en Fase 2 (el espectrograma lo codifica implícitamente) |

**Salidas generadas:**
- `ML/models/seismic_int8.tflite` — modelo cuantizado INT8 (pesos) / float32 (I/O)
- `ML/models/model.h` — arreglo C para incrustar en firmware ESP32
- `ML/models/norm_constants.h` — MEAN, STD y HANN_POWER para normalización en C++
- `ML/models/training_curves.png` — curvas de pérdida y AUC
- `ML/models/confusion_matrix.png` — matriz de confusión en test set


In [1]:
import os

# Forzar ejecución en CPU.
# Si tienes GPU y quieres usarla, comenta la siguiente línea:
os.environ["CUDA_VISIBLE_DEVICES"] = "-1"

import tensorflow as tf
print(f"TensorFlow {tf.__version__}")
print(f"GPUs visibles : {tf.config.list_physical_devices('GPU')}")
print(f"CPUs visibles : {tf.config.list_physical_devices('CPU')}")


I0000 00:00:1778679777.073419  202028 cpu_feature_guard.cc:227] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.


TensorFlow 2.21.0
GPUs visibles : []
CPUs visibles : [PhysicalDevice(name='/physical_device:CPU:0', device_type='CPU')]


## Celda 2 — Imports y constantes

Centraliza todos los hiperparámetros del experimento en un único lugar para
facilitar la experimentación sin tocar el resto del notebook.

| Constante | Valor | Descripción |
|-----------|-------|-------------|
| `INPUT_SHAPE` | `(65, 12, 3)` | Forma de cada espectrograma (frec, tiempo, canal) |
| `BATCH_SIZE` | 32 | Muestras por mini-batch |
| `MAX_EPOCHS` | 100 | Límite superior; EarlyStopping cortará antes |
| `LR_INIT` | 1e-3 | Tasa de aprendizaje inicial de Adam |
| `SEED` | 42 | Semilla global para reproducibilidad del split |


In [2]:
import numpy as np
from pathlib import Path
from tensorflow import keras
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report, confusion_matrix, ConfusionMatrixDisplay
import matplotlib
matplotlib.use('Agg')  # backend sin pantalla para nbconvert
import matplotlib.pyplot as plt

RUIDO_DIR   = Path("../Data_Labeling/Dataset/0_Ruido")
SISMO_DIR   = Path("../Data_Labeling/Dataset/1_Sismo")
MODEL_DIR   = Path("models"); MODEL_DIR.mkdir(exist_ok=True)

INPUT_SHAPE = (65, 11, 3)
BATCH_SIZE  = 32
MAX_EPOCHS  = 100
LR_INIT     = 1e-3
SEED        = 42


## Celda 4 — Carga del dataset

Carga todos los archivos `.npy` de ambas clases en arrays NumPy.
El orden de carga es reproducible (`sorted`) pero el split posterior mezcla
con semilla fija, por lo que el orden de carga no afecta el resultado.

**class_weight:** compensa el desbalance 1.64:1 (ruido:sismo) penalizando
más los errores sobre la clase minoritaria (sismo) durante el entrenamiento.
El factor es `n_ruido_train / n_sismo_train` para la clase 1; la clase 0 recibe peso 1.

**Split 70 / 15 / 15 estratificado:** mantiene la proporción 1.64:1 en
train, val y test, evitando que el test set tenga distribución diferente.


In [3]:
def load_dataset(ruido_dir, sismo_dir):
    X, y = [], []
    for npy in sorted(Path(ruido_dir).glob('*.npy')): X.append(np.load(npy)); y.append(0)
    for npy in sorted(Path(sismo_dir).glob('*.npy')): X.append(np.load(npy)); y.append(1)
    return np.stack(X, dtype=np.float32), np.array(y, dtype=np.int32)

X, y = load_dataset(RUIDO_DIR, SISMO_DIR)
print(f"Dataset cargado: {X.shape}  dtype={X.dtype}")
print(f"  Ruido (0): {np.sum(y==0)}   Sismo (1): {np.sum(y==1)}")

X_trv,  X_test,  y_trv,  y_test  = train_test_split(X,   y,   test_size=0.15,  random_state=SEED, stratify=y)
X_train, X_val,  y_train, y_val  = train_test_split(X_trv, y_trv, test_size=0.176, random_state=SEED, stratify=y_trv)

n0, n1 = int(np.sum(y_train==0)), int(np.sum(y_train==1))
class_weight = {0: 1.0, 1: round(n0 / n1, 4)}
print(f"\nTrain={len(y_train)}  Val={len(y_val)}  Test={len(y_test)}")
print(f"class_weight = {class_weight}")


Dataset cargado: (5288, 65, 11, 3)  dtype=float32
  Ruido (0): 3284   Sismo (1): 2004

Train=3703  Val=791  Test=794
class_weight = {0: 1.0, 1: 1.6393}


## Celda 6 — Normalización Z-score global

**¿Por qué normalizar?** Los valores raw de `log10(PSD + 1e-12)` oscilan
entre −12 (señal nula) y −2 (señal muy energética). Sin normalización,
la función de pérdida converge lento porque los gradientes son pequeños
fuera del rango [−1, 1].

**¿Por qué Z-score escalar (un solo MEAN y STD)?**
- Solo requiere 2 constantes en el firmware del ESP32 (2 floats = 8 bytes).
- El log-espectrograma tiene distribución aproximadamente normal, así que
  el Z-score global es casi tan efectivo como una normalización por bin.
- Se computa **únicamente sobre el training set** para evitar data leakage:
  si usáramos val o test para calcular la media, estaríamos "viendo" el
  futuro durante el entrenamiento.

**En el ESP32:** `x_norm = (log10(PSD + 1e-12) − MEAN) / STD`  
aplicado a cada uno de los 65 × 12 × 3 = 2 340 valores del espectrograma.


In [4]:
NORM_MEAN = float(X_train.mean())
NORM_STD  = float(X_train.std())
print(f"NORM_MEAN = {NORM_MEAN:.8f}")
print(f"NORM_STD  = {NORM_STD:.8f}")
print(f"Rango raw  : [{X_train.min():.2f}, {X_train.max():.2f}]")
X_n_tmp = (X_train - NORM_MEAN) / NORM_STD
print(f"Rango norm : [{X_n_tmp.min():.2f}, {X_n_tmp.max():.2f}]")
del X_n_tmp

X_train_n = (X_train - NORM_MEAN) / NORM_STD
X_val_n   = (X_val   - NORM_MEAN) / NORM_STD
X_test_n  = (X_test  - NORM_MEAN) / NORM_STD

np.savez(MODEL_DIR / "norm_params.npz",
         mean=np.float32(NORM_MEAN), std=np.float32(NORM_STD))
print("\nnorm_params.npz guardado")


NORM_MEAN = -9.83630657
NORM_STD  = 1.83801067
Rango raw  : [-12.00, -0.99]
Rango norm : [-1.18, 4.81]

norm_params.npz guardado


## Celda 8 — Arquitectura 1D-CNN

El espectrograma tiene forma `(65 frec, 11 tiempo, 3 canales)`.
Para una Conv1D necesitamos que la dimensión de **secuencia** sea el eje temporal:

```
(65, 11, 3)  →  Permute(2,1,3)  →  (11, 65, 3)  →  Reshape  →  (11, 195)
                                                                      ↑
                                               195 = 65 bins × 3 canales
```

Cada uno de los 11 pasos temporales (~0.33 s) tiene 195 features que
codifican el contenido frecuencial completo de los 3 ejes en ese instante.

| Capa | Salida | Parámetros |
|------|--------|------------|
| Input | (65, 11, 3) | — |
| Permute + Reshape | (11, 195) | 0 |
| Conv1D(32, k=3) + BN + ReLU | (11, 32) | 18 848 |
| MaxPool1D(2) | (5, 32) | 0 |
| Conv1D(64, k=3) + BN + ReLU | (5, 64) | 6 400 |
| MaxPool1D(2) | (2, 64) | 0 |
| GlobalAveragePooling1D | (64,) | 0 |
| Dropout(0.3) | (64,) | 0 |
| Dense(32, relu) | (32,) | 2 080 |
| Dense(1, sigmoid) | (1,) | 33 |
| **Total** | | **~27 500 params** |

FP32 ≈ 110 KB · INT8 ≈ 30 KB — bien por debajo del límite de 300 KB.


In [5]:
def build_model(input_shape=INPUT_SHAPE):
    inp = keras.Input(shape=input_shape)
    x   = keras.layers.Permute((2, 1, 3))(inp)        # (12, 65, 3)
    x   = keras.layers.Reshape((11, 65 * 3))(x)       # (11, 195)

    x = keras.layers.Conv1D(32, 3, padding='same')(x)
    x = keras.layers.BatchNormalization()(x)
    x = keras.layers.ReLU()(x)
    x = keras.layers.MaxPool1D(2)(x)                   # (5, 32)

    x = keras.layers.Conv1D(64, 3, padding='same')(x)
    x = keras.layers.BatchNormalization()(x)
    x = keras.layers.ReLU()(x)
    x = keras.layers.MaxPool1D(2)(x)                   # (2, 64)

    x = keras.layers.GlobalAveragePooling1D()(x)       # (64,)
    x = keras.layers.Dropout(0.3)(x)
    x = keras.layers.Dense(32, activation='relu')(x)
    out = keras.layers.Dense(1, activation='sigmoid')(x)
    return keras.Model(inp, out, name='seismic_1dcnn')

model = build_model()
model.summary()


Model: "seismic_1dcnn"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ input_layer (InputLayer)        │ (None, 65, 11, 3)      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ permute (Permute)               │ (None, 11, 65, 3)      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ reshape (Reshape)               │ (None, 11, 195)        │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv1d (Conv1D)                 │ (None, 11, 32)         │        18,752 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ batch_normalization             │ (None, 11, 32)         │           128 │
│ (BatchNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ re_lu (ReLU)                    │ (None, 11, 32)         │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling1d (MaxPooling1D)    │ (None, 5, 32)          │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv1d_1 (Conv1D)               │ (None, 5, 64)          │         6,208 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ batch_normalization_1           │ (None, 5, 64)          │           256 │
│ (BatchNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ re_lu_1 (ReLU)                  │ (None, 5, 64)          │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling1d_1 (MaxPooling1D)  │ (None, 2, 64)          │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ global_average_pooling1d        │ (None, 64)             │             0 │
│ (GlobalAveragePooling1D)        │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout (Dropout)               │ (None, 64)             │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ (None, 32)             │         2,080 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 1)              │            33 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 27,457 (107.25 KB)

 Trainable params: 27,265 (106.50 KB)

 Non-trainable params: 192 (768.00 B)

## Celda 10 — Entrenamiento

**Optimizador Adam:** adapta la tasa de aprendizaje por parámetro;
converge más rápido que SGD puro sin ajuste manual del scheduler.

**Métricas:**
- `accuracy` — orientativa (engañosa con clases desbalanceadas)
- `roc_auc` — área bajo la curva ROC; métrica principal de monitoreo

| Callback | Acción |
|----------|--------|
| `EarlyStopping(patience=15)` | Para si val_roc_auc no mejora 15 épocas |
| `ReduceLROnPlateau(patience=5)` | Divide LR × 0.3 si val_loss estanca |
| `ModelCheckpoint` | Guarda el mejor modelo según val_roc_auc |

`restore_best_weights=True` asegura que al final del entrenamiento el
modelo activo es el del mejor epoch, no el del último.


In [6]:
import tensorflow as tf

model.compile(
    optimizer=keras.optimizers.Adam(LR_INIT),
    loss='binary_crossentropy',
    metrics=['accuracy', keras.metrics.AUC(name='roc_auc')]
)

callbacks = [
    keras.callbacks.EarlyStopping(
        monitor='val_roc_auc', patience=15, mode='max', restore_best_weights=True),
    keras.callbacks.ReduceLROnPlateau(
        monitor='val_loss', factor=0.3, patience=5, min_lr=1e-6),
    keras.callbacks.ModelCheckpoint(
        MODEL_DIR / 'best_model.keras', monitor='val_roc_auc', mode='max', save_best_only=True)
]

history = model.fit(
    X_train_n, y_train,
    validation_data=(X_val_n, y_val),
    epochs=MAX_EPOCHS, batch_size=BATCH_SIZE,
    class_weight=class_weight,
    callbacks=callbacks,
    verbose=1
)

fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(12, 4))
ax1.plot(history.history['loss'], label='train')
ax1.plot(history.history['val_loss'], label='val')
ax1.set(title='Loss', xlabel='epoch'); ax1.legend()
ax2.plot(history.history['roc_auc'], label='train')
ax2.plot(history.history['val_roc_auc'], label='val')
ax2.set(title='ROC-AUC', xlabel='epoch'); ax2.legend()
plt.tight_layout()
plt.savefig(MODEL_DIR / 'training_curves.png', dpi=100)
plt.close()
print("training_curves.png guardado")


Epoch 1/100


  1/116 ━━━━━━━━━━━━━━━━━━━━ 4:16 2s/step - accuracy: 0.5000 - loss: 1.0571 - roc_auc: 0.5875

 13/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.5539 - loss: 0.9132 - roc_auc: 0.6583 

 29/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.6206 - loss: 0.8370 - roc_auc: 0.6925

 44/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.6486 - loss: 0.7989 - roc_auc: 0.7132

 59/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.6652 - loss: 0.7745 - roc_auc: 0.7275

 74/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.6759 - loss: 0.7553 - roc_auc: 0.7385

 89/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.6850 - loss: 0.7394 - roc_auc: 0.7476

103/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.6920 - loss: 0.7272 - roc_auc: 0.7546

116/116 ━━━━━━━━━━━━━━━━━━━━ 3s 8ms/step - accuracy: 0.7313 - loss: 0.6492 - roc_auc: 0.7999 - val_accuracy: 0.7636 - val_loss: 0.4971 - val_roc_auc: 0.8325 - learning_rate: 0.0010


Epoch 2/100


  1/116 ━━━━━━━━━━━━━━━━━━━━ 4s 36ms/step - accuracy: 0.6562 - loss: 0.6606 - roc_auc: 0.7157

 16/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.6994 - loss: 0.6650 - roc_auc: 0.7766 

 30/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7121 - loss: 0.6489 - roc_auc: 0.7884

 45/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7229 - loss: 0.6370 - roc_auc: 0.7965

 59/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7301 - loss: 0.6295 - roc_auc: 0.8012

 74/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7366 - loss: 0.6229 - roc_auc: 0.8058

 88/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7413 - loss: 0.6185 - roc_auc: 0.8083

102/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7445 - loss: 0.6150 - roc_auc: 0.8105

116/116 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.7645 - loss: 0.5906 - roc_auc: 0.8247 - val_accuracy: 0.7901 - val_loss: 0.4525 - val_roc_auc: 0.8526 - learning_rate: 0.0010


Epoch 3/100


  1/116 ━━━━━━━━━━━━━━━━━━━━ 2s 26ms/step - accuracy: 0.7812 - loss: 0.5878 - roc_auc: 0.7917

 14/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8063 - loss: 0.5679 - roc_auc: 0.8433 

 27/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7942 - loss: 0.5700 - roc_auc: 0.8426

 40/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7861 - loss: 0.5737 - roc_auc: 0.8402

 53/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7813 - loss: 0.5759 - roc_auc: 0.8391

 67/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7788 - loss: 0.5770 - roc_auc: 0.8386

 81/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7769 - loss: 0.5773 - roc_auc: 0.8383

 95/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7764 - loss: 0.5763 - roc_auc: 0.8388

107/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7764 - loss: 0.5757 - roc_auc: 0.8391

116/116 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.7769 - loss: 0.5680 - roc_auc: 0.8385 - val_accuracy: 0.7699 - val_loss: 0.4241 - val_roc_auc: 0.8584 - learning_rate: 0.0010


Epoch 4/100


  1/116 ━━━━━━━━━━━━━━━━━━━━ 3s 29ms/step - accuracy: 0.8125 - loss: 0.5412 - roc_auc: 0.8765

 16/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7969 - loss: 0.5465 - roc_auc: 0.8639 

 30/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7918 - loss: 0.5532 - roc_auc: 0.8594

 44/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7861 - loss: 0.5574 - roc_auc: 0.8556

 58/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7843 - loss: 0.5571 - roc_auc: 0.8550

 72/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7832 - loss: 0.5564 - roc_auc: 0.8545

 87/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7823 - loss: 0.5559 - roc_auc: 0.8541

101/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7814 - loss: 0.5556 - roc_auc: 0.8536

114/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7811 - loss: 0.5549 - roc_auc: 0.8534

116/116 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.7823 - loss: 0.5468 - roc_auc: 0.8533 - val_accuracy: 0.7863 - val_loss: 0.4331 - val_roc_auc: 0.8516 - learning_rate: 0.0010


Epoch 5/100


  1/116 ━━━━━━━━━━━━━━━━━━━━ 2s 24ms/step - accuracy: 0.8438 - loss: 0.6612 - roc_auc: 0.7794

 15/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8013 - loss: 0.5836 - roc_auc: 0.8395 

 30/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7889 - loss: 0.5794 - roc_auc: 0.8412

 45/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7810 - loss: 0.5769 - roc_auc: 0.8417

 60/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7774 - loss: 0.5752 - roc_auc: 0.8414

 74/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7767 - loss: 0.5733 - roc_auc: 0.8418

 89/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7762 - loss: 0.5724 - roc_auc: 0.8417

104/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7764 - loss: 0.5702 - roc_auc: 0.8426

116/116 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.7818 - loss: 0.5522 - roc_auc: 0.8502 - val_accuracy: 0.7724 - val_loss: 0.4023 - val_roc_auc: 0.8762 - learning_rate: 0.0010


Epoch 6/100


  1/116 ━━━━━━━━━━━━━━━━━━━━ 2s 23ms/step - accuracy: 0.8750 - loss: 0.4066 - roc_auc: 0.9156

 15/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8069 - loss: 0.5435 - roc_auc: 0.8432 

 30/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7917 - loss: 0.5572 - roc_auc: 0.8464

 44/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7875 - loss: 0.5603 - roc_auc: 0.8475

 58/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7848 - loss: 0.5600 - roc_auc: 0.8489

 73/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7843 - loss: 0.5575 - roc_auc: 0.8509

 88/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7853 - loss: 0.5543 - roc_auc: 0.8531

103/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7860 - loss: 0.5517 - roc_auc: 0.8546

116/116 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.7921 - loss: 0.5315 - roc_auc: 0.8649 - val_accuracy: 0.7674 - val_loss: 0.5743 - val_roc_auc: 0.8700 - learning_rate: 0.0010


Epoch 7/100


  1/116 ━━━━━━━━━━━━━━━━━━━━ 2s 23ms/step - accuracy: 0.7188 - loss: 0.5709 - roc_auc: 0.8117

 16/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7671 - loss: 0.5743 - roc_auc: 0.8220 

 30/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7805 - loss: 0.5559 - roc_auc: 0.8399

 45/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7838 - loss: 0.5508 - roc_auc: 0.8464

 59/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7859 - loss: 0.5474 - roc_auc: 0.8498

 74/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7887 - loss: 0.5433 - roc_auc: 0.8534

 88/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7912 - loss: 0.5401 - roc_auc: 0.8563

102/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7931 - loss: 0.5376 - roc_auc: 0.8584

116/116 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.8045 - loss: 0.5145 - roc_auc: 0.8741 - val_accuracy: 0.7206 - val_loss: 0.5618 - val_roc_auc: 0.8551 - learning_rate: 0.0010


Epoch 8/100


  1/116 ━━━━━━━━━━━━━━━━━━━━ 2s 26ms/step - accuracy: 0.9062 - loss: 0.3897 - roc_auc: 0.9570

 15/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8010 - loss: 0.5487 - roc_auc: 0.8703 

 30/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7849 - loss: 0.5498 - roc_auc: 0.8651

 44/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7830 - loss: 0.5467 - roc_auc: 0.8646

 58/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7841 - loss: 0.5433 - roc_auc: 0.8651

 72/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7872 - loss: 0.5395 - roc_auc: 0.8663

 86/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7901 - loss: 0.5358 - roc_auc: 0.8675

100/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7922 - loss: 0.5327 - roc_auc: 0.8685

114/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7936 - loss: 0.5309 - roc_auc: 0.8691

116/116 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.8021 - loss: 0.5236 - roc_auc: 0.8721 - val_accuracy: 0.6574 - val_loss: 0.8828 - val_roc_auc: 0.6258 - learning_rate: 0.0010


Epoch 9/100


  1/116 ━━━━━━━━━━━━━━━━━━━━ 2s 23ms/step - accuracy: 0.6250 - loss: 0.7680 - roc_auc: 0.7004

 15/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7624 - loss: 0.6369 - roc_auc: 0.8108 

 30/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7682 - loss: 0.6146 - roc_auc: 0.8256

 44/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7695 - loss: 0.6056 - roc_auc: 0.8300

 58/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7721 - loss: 0.5972 - roc_auc: 0.8338

 72/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7745 - loss: 0.5903 - roc_auc: 0.8365

 86/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7756 - loss: 0.5850 - roc_auc: 0.8386

101/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7767 - loss: 0.5798 - roc_auc: 0.8406

115/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7782 - loss: 0.5749 - roc_auc: 0.8425

116/116 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.7921 - loss: 0.5339 - roc_auc: 0.8602 - val_accuracy: 0.7788 - val_loss: 0.5679 - val_roc_auc: 0.8759 - learning_rate: 0.0010


Epoch 10/100


  1/116 ━━━━━━━━━━━━━━━━━━━━ 2s 23ms/step - accuracy: 0.8125 - loss: 0.5051 - roc_auc: 0.8909

 16/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8356 - loss: 0.4782 - roc_auc: 0.9039 

 30/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8171 - loss: 0.4977 - roc_auc: 0.8898

 44/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8056 - loss: 0.5058 - roc_auc: 0.8830

 58/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8016 - loss: 0.5086 - roc_auc: 0.8799

 72/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8006 - loss: 0.5093 - roc_auc: 0.8782

 86/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7999 - loss: 0.5097 - roc_auc: 0.8768

100/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7994 - loss: 0.5105 - roc_auc: 0.8755

114/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7991 - loss: 0.5113 - roc_auc: 0.8746

116/116 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.7988 - loss: 0.5141 - roc_auc: 0.8707 - val_accuracy: 0.7598 - val_loss: 0.5214 - val_roc_auc: 0.8747 - learning_rate: 0.0010


Epoch 11/100


  1/116 ━━━━━━━━━━━━━━━━━━━━ 2s 24ms/step - accuracy: 0.7500 - loss: 0.3684 - roc_auc: 0.9903

 16/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.7805 - loss: 0.4386 - roc_auc: 0.9135 

 31/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8005 - loss: 0.4562 - roc_auc: 0.9034

 44/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8068 - loss: 0.4638 - roc_auc: 0.8996

 58/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8083 - loss: 0.4686 - roc_auc: 0.8972

 70/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8095 - loss: 0.4707 - roc_auc: 0.8962

 83/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8103 - loss: 0.4724 - roc_auc: 0.8952

 97/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8113 - loss: 0.4737 - roc_auc: 0.8944

110/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8119 - loss: 0.4751 - roc_auc: 0.8936

116/116 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.8153 - loss: 0.4864 - roc_auc: 0.8881 - val_accuracy: 0.8420 - val_loss: 0.3859 - val_roc_auc: 0.9148 - learning_rate: 3.0000e-04


Epoch 12/100


  1/116 ━━━━━━━━━━━━━━━━━━━━ 2s 24ms/step - accuracy: 0.9062 - loss: 0.3620 - roc_auc: 0.9372

 15/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8859 - loss: 0.4077 - roc_auc: 0.9288 

 27/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8713 - loss: 0.4274 - roc_auc: 0.9195

 40/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8641 - loss: 0.4356 - roc_auc: 0.9163

 53/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8604 - loss: 0.4410 - roc_auc: 0.9140

 67/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8571 - loss: 0.4456 - roc_auc: 0.9116

 81/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8545 - loss: 0.4484 - roc_auc: 0.9099

 95/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8521 - loss: 0.4505 - roc_auc: 0.9087

109/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8501 - loss: 0.4523 - roc_auc: 0.9074

116/116 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.8323 - loss: 0.4727 - roc_auc: 0.8950 - val_accuracy: 0.8496 - val_loss: 0.3407 - val_roc_auc: 0.9151 - learning_rate: 3.0000e-04


Epoch 13/100


  1/116 ━━━━━━━━━━━━━━━━━━━━ 2s 21ms/step - accuracy: 0.9375 - loss: 0.3335 - roc_auc: 0.9886

 15/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8546 - loss: 0.4391 - roc_auc: 0.9111 

 29/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8380 - loss: 0.4553 - roc_auc: 0.9037

 43/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8335 - loss: 0.4564 - roc_auc: 0.9029

 58/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8314 - loss: 0.4578 - roc_auc: 0.9020

 73/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8310 - loss: 0.4576 - roc_auc: 0.9018

 87/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8315 - loss: 0.4573 - roc_auc: 0.9016

102/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8320 - loss: 0.4571 - roc_auc: 0.9016

116/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8321 - loss: 0.4570 - roc_auc: 0.9017

116/116 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.8328 - loss: 0.4557 - roc_auc: 0.9023 - val_accuracy: 0.8584 - val_loss: 0.3322 - val_roc_auc: 0.9049 - learning_rate: 3.0000e-04


Epoch 14/100


  1/116 ━━━━━━━━━━━━━━━━━━━━ 2s 23ms/step - accuracy: 0.8750 - loss: 0.4303 - roc_auc: 0.9412

 16/116 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.8670 - loss: 0.4038 - roc_auc: 0.9297 

 29/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8519 - loss: 0.4194 - roc_auc: 0.9230

 44/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8440 - loss: 0.4312 - roc_auc: 0.9179

 58/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8410 - loss: 0.4351 - roc_auc: 0.9159

 73/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8403 - loss: 0.4364 - roc_auc: 0.9151

 88/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8407 - loss: 0.4366 - roc_auc: 0.9149

102/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8403 - loss: 0.4372 - roc_auc: 0.9145

116/116 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.8358 - loss: 0.4459 - roc_auc: 0.9086 - val_accuracy: 0.8458 - val_loss: 0.3408 - val_roc_auc: 0.9122 - learning_rate: 3.0000e-04


Epoch 15/100


  1/116 ━━━━━━━━━━━━━━━━━━━━ 3s 26ms/step - accuracy: 0.8125 - loss: 0.4120 - roc_auc: 0.8802

 16/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8352 - loss: 0.4385 - roc_auc: 0.9138 

 31/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8384 - loss: 0.4436 - roc_auc: 0.9131

 44/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8406 - loss: 0.4410 - roc_auc: 0.9148

 58/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8386 - loss: 0.4410 - roc_auc: 0.9149

 72/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8391 - loss: 0.4387 - roc_auc: 0.9160

 87/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8398 - loss: 0.4369 - roc_auc: 0.9169

101/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8407 - loss: 0.4353 - roc_auc: 0.9176

116/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8415 - loss: 0.4337 - roc_auc: 0.9181

116/116 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.8439 - loss: 0.4246 - roc_auc: 0.9201 - val_accuracy: 0.8609 - val_loss: 0.3213 - val_roc_auc: 0.9169 - learning_rate: 3.0000e-04


Epoch 16/100


  1/116 ━━━━━━━━━━━━━━━━━━━━ 2s 25ms/step - accuracy: 0.9062 - loss: 0.3941 - roc_auc: 0.9802

 16/116 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.8384 - loss: 0.4317 - roc_auc: 0.9238 

 30/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8347 - loss: 0.4330 - roc_auc: 0.9193

 44/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8363 - loss: 0.4300 - roc_auc: 0.9189

 58/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8382 - loss: 0.4287 - roc_auc: 0.9187

 73/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8392 - loss: 0.4280 - roc_auc: 0.9187

 88/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8397 - loss: 0.4286 - roc_auc: 0.9182

102/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8402 - loss: 0.4287 - roc_auc: 0.9181

116/116 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.8461 - loss: 0.4191 - roc_auc: 0.9215 - val_accuracy: 0.8420 - val_loss: 0.3288 - val_roc_auc: 0.9152 - learning_rate: 3.0000e-04


Epoch 17/100


  1/116 ━━━━━━━━━━━━━━━━━━━━ 2s 23ms/step - accuracy: 0.8438 - loss: 0.5491 - roc_auc: 0.8227

 15/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8601 - loss: 0.4505 - roc_auc: 0.9082 

 30/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8644 - loss: 0.4323 - roc_auc: 0.9155

 44/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8622 - loss: 0.4258 - roc_auc: 0.9188

 57/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8608 - loss: 0.4216 - roc_auc: 0.9208

 72/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8598 - loss: 0.4177 - roc_auc: 0.9226

 86/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8588 - loss: 0.4157 - roc_auc: 0.9235

101/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8584 - loss: 0.4140 - roc_auc: 0.9243

116/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8581 - loss: 0.4128 - roc_auc: 0.9248

116/116 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.8561 - loss: 0.4055 - roc_auc: 0.9286 - val_accuracy: 0.8546 - val_loss: 0.3510 - val_roc_auc: 0.8876 - learning_rate: 3.0000e-04


Epoch 18/100


  1/116 ━━━━━━━━━━━━━━━━━━━━ 2s 23ms/step - accuracy: 0.9062 - loss: 0.2651 - roc_auc: 0.9827

 16/116 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.9006 - loss: 0.3411 - roc_auc: 0.9485 

 30/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8921 - loss: 0.3505 - roc_auc: 0.9466

 44/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8871 - loss: 0.3561 - roc_auc: 0.9452

 58/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8848 - loss: 0.3593 - roc_auc: 0.9447

 72/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8817 - loss: 0.3651 - roc_auc: 0.9431

 87/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8791 - loss: 0.3700 - roc_auc: 0.9417

101/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8774 - loss: 0.3735 - roc_auc: 0.9405

116/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8758 - loss: 0.3769 - roc_auc: 0.9393

116/116 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.8644 - loss: 0.4026 - roc_auc: 0.9302 - val_accuracy: 0.8104 - val_loss: 0.3735 - val_roc_auc: 0.9150 - learning_rate: 3.0000e-04


Epoch 19/100


  1/116 ━━━━━━━━━━━━━━━━━━━━ 2s 23ms/step - accuracy: 0.7500 - loss: 0.4590 - roc_auc: 0.8863

 16/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8224 - loss: 0.4143 - roc_auc: 0.9268 

 31/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8396 - loss: 0.4023 - roc_auc: 0.9310

 45/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8454 - loss: 0.3970 - roc_auc: 0.9330

 60/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8486 - loss: 0.3942 - roc_auc: 0.9340

 73/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8511 - loss: 0.3921 - roc_auc: 0.9346

 87/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8533 - loss: 0.3902 - roc_auc: 0.9352

101/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8554 - loss: 0.3882 - roc_auc: 0.9359

115/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8570 - loss: 0.3866 - roc_auc: 0.9363

116/116 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.8693 - loss: 0.3748 - roc_auc: 0.9394 - val_accuracy: 0.8445 - val_loss: 0.3182 - val_roc_auc: 0.9204 - learning_rate: 3.0000e-04


Epoch 20/100


  1/116 ━━━━━━━━━━━━━━━━━━━━ 2s 25ms/step - accuracy: 0.9062 - loss: 0.3983 - roc_auc: 0.9453

 15/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8531 - loss: 0.4331 - roc_auc: 0.9205 

 28/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8469 - loss: 0.4369 - roc_auc: 0.9186

 42/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8512 - loss: 0.4269 - roc_auc: 0.9216

 56/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8564 - loss: 0.4153 - roc_auc: 0.9254

 69/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8595 - loss: 0.4084 - roc_auc: 0.9278

 82/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8618 - loss: 0.4036 - roc_auc: 0.9295

 94/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8631 - loss: 0.4009 - roc_auc: 0.9304

107/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8641 - loss: 0.3984 - roc_auc: 0.9313

116/116 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.8758 - loss: 0.3751 - roc_auc: 0.9394 - val_accuracy: 0.8559 - val_loss: 0.3136 - val_roc_auc: 0.9151 - learning_rate: 3.0000e-04


Epoch 21/100


  1/116 ━━━━━━━━━━━━━━━━━━━━ 3s 26ms/step - accuracy: 0.8750 - loss: 0.3469 - roc_auc: 0.9676

 15/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8789 - loss: 0.3449 - roc_auc: 0.9537 

 28/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8769 - loss: 0.3430 - roc_auc: 0.9516

 42/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8727 - loss: 0.3499 - roc_auc: 0.9480

 56/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8712 - loss: 0.3545 - roc_auc: 0.9459

 70/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8705 - loss: 0.3573 - roc_auc: 0.9446

 84/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8707 - loss: 0.3582 - roc_auc: 0.9441

 99/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8713 - loss: 0.3582 - roc_auc: 0.9439

113/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8722 - loss: 0.3571 - roc_auc: 0.9443

116/116 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.8809 - loss: 0.3454 - roc_auc: 0.9486 - val_accuracy: 0.8635 - val_loss: 0.3619 - val_roc_auc: 0.8942 - learning_rate: 3.0000e-04


Epoch 22/100


  1/116 ━━━━━━━━━━━━━━━━━━━━ 2s 23ms/step - accuracy: 0.9375 - loss: 0.2315 - roc_auc: 1.0000

 16/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8981 - loss: 0.3129 - roc_auc: 0.9662 

 30/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8986 - loss: 0.3186 - roc_auc: 0.9614

 43/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8984 - loss: 0.3206 - roc_auc: 0.9600

 57/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8988 - loss: 0.3196 - roc_auc: 0.9597

 70/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8995 - loss: 0.3191 - roc_auc: 0.9595

 84/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8993 - loss: 0.3202 - roc_auc: 0.9589

 99/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8991 - loss: 0.3215 - roc_auc: 0.9583

114/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8992 - loss: 0.3223 - roc_auc: 0.9579

116/116 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.8990 - loss: 0.3292 - roc_auc: 0.9543 - val_accuracy: 0.8546 - val_loss: 0.4192 - val_roc_auc: 0.8869 - learning_rate: 3.0000e-04


Epoch 23/100


  1/116 ━━━━━━━━━━━━━━━━━━━━ 2s 23ms/step - accuracy: 0.9688 - loss: 0.1937 - roc_auc: 0.9909

 15/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9261 - loss: 0.2864 - roc_auc: 0.9682 

 30/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9081 - loss: 0.3135 - roc_auc: 0.9605

 44/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9031 - loss: 0.3175 - roc_auc: 0.9594

 59/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9015 - loss: 0.3185 - roc_auc: 0.9591

 73/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9012 - loss: 0.3178 - roc_auc: 0.9591

 87/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9012 - loss: 0.3166 - roc_auc: 0.9593

101/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9010 - loss: 0.3160 - roc_auc: 0.9595

115/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9011 - loss: 0.3153 - roc_auc: 0.9596

116/116 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.9009 - loss: 0.3107 - roc_auc: 0.9608 - val_accuracy: 0.7863 - val_loss: 0.3738 - val_roc_auc: 0.9087 - learning_rate: 3.0000e-04


Epoch 24/100


  1/116 ━━━━━━━━━━━━━━━━━━━━ 2s 22ms/step - accuracy: 0.7812 - loss: 0.4072 - roc_auc: 0.9481

 15/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8861 - loss: 0.3091 - roc_auc: 0.9621 

 29/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8935 - loss: 0.3068 - roc_auc: 0.9618

 44/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8923 - loss: 0.3107 - roc_auc: 0.9601

 58/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8926 - loss: 0.3101 - roc_auc: 0.9602

 72/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8932 - loss: 0.3089 - roc_auc: 0.9604

 86/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8937 - loss: 0.3085 - roc_auc: 0.9605

100/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8939 - loss: 0.3082 - roc_auc: 0.9606

115/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8944 - loss: 0.3073 - roc_auc: 0.9608

116/116 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.8990 - loss: 0.3006 - roc_auc: 0.9628 - val_accuracy: 0.8281 - val_loss: 0.5002 - val_roc_auc: 0.8734 - learning_rate: 3.0000e-04


Epoch 25/100


  1/116 ━━━━━━━━━━━━━━━━━━━━ 2s 24ms/step - accuracy: 0.8750 - loss: 0.2488 - roc_auc: 0.9773

 15/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8886 - loss: 0.2906 - roc_auc: 0.9671 

 30/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8859 - loss: 0.3028 - roc_auc: 0.9635

 45/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8874 - loss: 0.3043 - roc_auc: 0.9630

 60/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8900 - loss: 0.3021 - roc_auc: 0.9635

 74/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8925 - loss: 0.2992 - roc_auc: 0.9641

 89/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8942 - loss: 0.2976 - roc_auc: 0.9644

102/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8951 - loss: 0.2972 - roc_auc: 0.9644

116/116 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.9006 - loss: 0.2983 - roc_auc: 0.9632 - val_accuracy: 0.7762 - val_loss: 0.5160 - val_roc_auc: 0.9103 - learning_rate: 3.0000e-04


Epoch 26/100


  1/116 ━━━━━━━━━━━━━━━━━━━━ 2s 22ms/step - accuracy: 0.9062 - loss: 0.2845 - roc_auc: 0.9654

 14/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9118 - loss: 0.2828 - roc_auc: 0.9686 

 29/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9116 - loss: 0.2865 - roc_auc: 0.9673

 43/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9111 - loss: 0.2855 - roc_auc: 0.9672

 57/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9123 - loss: 0.2832 - roc_auc: 0.9676

 71/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9133 - loss: 0.2800 - roc_auc: 0.9683

 85/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9140 - loss: 0.2775 - roc_auc: 0.9689

100/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9142 - loss: 0.2759 - roc_auc: 0.9692

113/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9144 - loss: 0.2744 - roc_auc: 0.9695

116/116 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.9147 - loss: 0.2664 - roc_auc: 0.9709 - val_accuracy: 0.8255 - val_loss: 0.3603 - val_roc_auc: 0.9155 - learning_rate: 9.0000e-05


Epoch 27/100


  1/116 ━━━━━━━━━━━━━━━━━━━━ 2s 23ms/step - accuracy: 0.8125 - loss: 0.3043 - roc_auc: 0.9630

 16/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9006 - loss: 0.2446 - roc_auc: 0.9779 

 30/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9071 - loss: 0.2445 - roc_auc: 0.9782

 45/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9108 - loss: 0.2495 - roc_auc: 0.9770

 59/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9134 - loss: 0.2496 - roc_auc: 0.9768

 74/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9147 - loss: 0.2507 - roc_auc: 0.9764

 88/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9154 - loss: 0.2520 - roc_auc: 0.9759

103/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9162 - loss: 0.2525 - roc_auc: 0.9756

116/116 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.9201 - loss: 0.2534 - roc_auc: 0.9745 - val_accuracy: 0.8622 - val_loss: 0.3495 - val_roc_auc: 0.9065 - learning_rate: 9.0000e-05


Epoch 28/100


  1/116 ━━━━━━━━━━━━━━━━━━━━ 3s 27ms/step - accuracy: 0.8750 - loss: 0.4234 - roc_auc: 0.9668

 16/116 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.9076 - loss: 0.2868 - roc_auc: 0.9660 

 30/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9124 - loss: 0.2739 - roc_auc: 0.9687

 44/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9130 - loss: 0.2698 - roc_auc: 0.9699

 59/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9142 - loss: 0.2653 - roc_auc: 0.9711

 73/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9150 - loss: 0.2628 - roc_auc: 0.9718

 88/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9158 - loss: 0.2610 - roc_auc: 0.9722

102/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9165 - loss: 0.2591 - roc_auc: 0.9726

114/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9171 - loss: 0.2574 - roc_auc: 0.9730

116/116 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.9244 - loss: 0.2390 - roc_auc: 0.9769 - val_accuracy: 0.8546 - val_loss: 0.3403 - val_roc_auc: 0.9171 - learning_rate: 9.0000e-05


Epoch 29/100


  1/116 ━━━━━━━━━━━━━━━━━━━━ 3s 26ms/step - accuracy: 0.9062 - loss: 0.2849 - roc_auc: 0.9833

 15/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9330 - loss: 0.2267 - roc_auc: 0.9811 

 29/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9339 - loss: 0.2272 - roc_auc: 0.9804

 42/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9323 - loss: 0.2304 - roc_auc: 0.9797

 55/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9316 - loss: 0.2314 - roc_auc: 0.9794

 65/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9309 - loss: 0.2320 - roc_auc: 0.9793

 77/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9298 - loss: 0.2329 - roc_auc: 0.9791

 89/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9288 - loss: 0.2343 - roc_auc: 0.9788

101/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9280 - loss: 0.2352 - roc_auc: 0.9786

113/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9275 - loss: 0.2358 - roc_auc: 0.9784

116/116 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.9230 - loss: 0.2414 - roc_auc: 0.9767 - val_accuracy: 0.8546 - val_loss: 0.3422 - val_roc_auc: 0.9162 - learning_rate: 9.0000e-05


Epoch 30/100


  1/116 ━━━━━━━━━━━━━━━━━━━━ 2s 23ms/step - accuracy: 0.9688 - loss: 0.1743 - roc_auc: 0.9952

 16/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9390 - loss: 0.2181 - roc_auc: 0.9827 

 31/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9325 - loss: 0.2255 - roc_auc: 0.9811

 46/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9299 - loss: 0.2307 - roc_auc: 0.9797

 60/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9290 - loss: 0.2327 - roc_auc: 0.9791

 74/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9288 - loss: 0.2330 - roc_auc: 0.9789

 88/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9287 - loss: 0.2332 - roc_auc: 0.9789

102/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9287 - loss: 0.2330 - roc_auc: 0.9789

116/116 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.9274 - loss: 0.2352 - roc_auc: 0.9783 - val_accuracy: 0.8470 - val_loss: 0.3430 - val_roc_auc: 0.9186 - learning_rate: 9.0000e-05


Epoch 31/100


  1/116 ━━━━━━━━━━━━━━━━━━━━ 2s 24ms/step - accuracy: 0.9688 - loss: 0.1840 - roc_auc: 0.9961

 15/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9481 - loss: 0.1869 - roc_auc: 0.9895 

 30/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9468 - loss: 0.1915 - roc_auc: 0.9878

 43/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9440 - loss: 0.1984 - roc_auc: 0.9861

 57/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9415 - loss: 0.2029 - roc_auc: 0.9851

 72/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9403 - loss: 0.2043 - roc_auc: 0.9848

 87/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9392 - loss: 0.2056 - roc_auc: 0.9844

102/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9382 - loss: 0.2073 - roc_auc: 0.9840

116/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9374 - loss: 0.2085 - roc_auc: 0.9836

116/116 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.9322 - loss: 0.2178 - roc_auc: 0.9812 - val_accuracy: 0.8458 - val_loss: 0.3474 - val_roc_auc: 0.9174 - learning_rate: 2.7000e-05


Epoch 32/100


  1/116 ━━━━━━━━━━━━━━━━━━━━ 2s 22ms/step - accuracy: 0.9688 - loss: 0.1534 - roc_auc: 0.9955

 15/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9331 - loss: 0.2115 - roc_auc: 0.9837 

 29/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9310 - loss: 0.2190 - roc_auc: 0.9818

 44/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9328 - loss: 0.2162 - roc_auc: 0.9822

 58/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9334 - loss: 0.2154 - roc_auc: 0.9822

 73/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9331 - loss: 0.2159 - roc_auc: 0.9820

 87/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9332 - loss: 0.2162 - roc_auc: 0.9819

102/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9338 - loss: 0.2157 - roc_auc: 0.9820

116/116 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.9395 - loss: 0.2098 - roc_auc: 0.9832 - val_accuracy: 0.8521 - val_loss: 0.3454 - val_roc_auc: 0.9175 - learning_rate: 2.7000e-05


Epoch 33/100


  1/116 ━━━━━━━━━━━━━━━━━━━━ 3s 33ms/step - accuracy: 0.9375 - loss: 0.2039 - roc_auc: 0.9792

 15/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9373 - loss: 0.2037 - roc_auc: 0.9858 

 29/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9346 - loss: 0.2056 - roc_auc: 0.9849

 43/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9365 - loss: 0.2032 - roc_auc: 0.9852

 57/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9369 - loss: 0.2033 - roc_auc: 0.9849

 71/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9363 - loss: 0.2049 - roc_auc: 0.9844

 85/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9358 - loss: 0.2057 - roc_auc: 0.9841

 99/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9358 - loss: 0.2061 - roc_auc: 0.9840

113/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9358 - loss: 0.2066 - roc_auc: 0.9838

116/116 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.9346 - loss: 0.2145 - roc_auc: 0.9816 - val_accuracy: 0.8521 - val_loss: 0.3480 - val_roc_auc: 0.9197 - learning_rate: 2.7000e-05


Epoch 34/100


  1/116 ━━━━━━━━━━━━━━━━━━━━ 2s 24ms/step - accuracy: 0.9688 - loss: 0.1406 - roc_auc: 1.0000

 15/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9318 - loss: 0.2117 - roc_auc: 0.9814 

 29/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9274 - loss: 0.2141 - roc_auc: 0.9816

 43/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9295 - loss: 0.2106 - roc_auc: 0.9826

 57/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9318 - loss: 0.2071 - roc_auc: 0.9834

 71/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9334 - loss: 0.2056 - roc_auc: 0.9837

 85/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9343 - loss: 0.2055 - roc_auc: 0.9838

100/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9348 - loss: 0.2063 - roc_auc: 0.9836

114/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9351 - loss: 0.2072 - roc_auc: 0.9834

116/116 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.9365 - loss: 0.2148 - roc_auc: 0.9820 - val_accuracy: 0.8394 - val_loss: 0.3640 - val_roc_auc: 0.9204 - learning_rate: 2.7000e-05


Epoch 35/100


  1/116 ━━━━━━━━━━━━━━━━━━━━ 2s 25ms/step - accuracy: 0.9375 - loss: 0.1962 - roc_auc: 0.9922

 14/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9266 - loss: 0.2106 - roc_auc: 0.9857 

 28/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9238 - loss: 0.2162 - roc_auc: 0.9832

 42/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9246 - loss: 0.2160 - roc_auc: 0.9828

 56/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9263 - loss: 0.2134 - roc_auc: 0.9832

 69/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9278 - loss: 0.2123 - roc_auc: 0.9832

 83/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9292 - loss: 0.2114 - roc_auc: 0.9833

 97/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9300 - loss: 0.2107 - roc_auc: 0.9833

111/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9306 - loss: 0.2103 - roc_auc: 0.9833

116/116 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.9328 - loss: 0.2101 - roc_auc: 0.9829 - val_accuracy: 0.8369 - val_loss: 0.3696 - val_roc_auc: 0.9194 - learning_rate: 2.7000e-05


Epoch 36/100


  1/116 ━━━━━━━━━━━━━━━━━━━━ 2s 22ms/step - accuracy: 0.8750 - loss: 0.3349 - roc_auc: 0.9734

 15/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9187 - loss: 0.2598 - roc_auc: 0.9759 

 29/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9222 - loss: 0.2516 - roc_auc: 0.9764

 43/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9239 - loss: 0.2450 - roc_auc: 0.9771

 57/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9253 - loss: 0.2403 - roc_auc: 0.9777

 71/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9274 - loss: 0.2353 - roc_auc: 0.9785

 84/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9286 - loss: 0.2322 - roc_auc: 0.9789

 99/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9293 - loss: 0.2294 - roc_auc: 0.9793

113/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9298 - loss: 0.2276 - roc_auc: 0.9796

116/116 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.9330 - loss: 0.2172 - roc_auc: 0.9811 - val_accuracy: 0.8508 - val_loss: 0.3535 - val_roc_auc: 0.9183 - learning_rate: 8.1000e-06


Epoch 37/100


  1/116 ━━━━━━━━━━━━━━━━━━━━ 2s 23ms/step - accuracy: 0.9375 - loss: 0.1529 - roc_auc: 1.0000

 16/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9518 - loss: 0.1968 - roc_auc: 0.9870 

 29/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9488 - loss: 0.2012 - roc_auc: 0.9854

 42/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9480 - loss: 0.2005 - roc_auc: 0.9853

 54/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9462 - loss: 0.2028 - roc_auc: 0.9849

 66/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9452 - loss: 0.2044 - roc_auc: 0.9846

 79/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9449 - loss: 0.2044 - roc_auc: 0.9846

 92/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9448 - loss: 0.2045 - roc_auc: 0.9846

106/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9445 - loss: 0.2051 - roc_auc: 0.9844

116/116 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.9409 - loss: 0.2088 - roc_auc: 0.9835 - val_accuracy: 0.8458 - val_loss: 0.3548 - val_roc_auc: 0.9191 - learning_rate: 8.1000e-06


Epoch 38/100


  1/116 ━━━━━━━━━━━━━━━━━━━━ 2s 24ms/step - accuracy: 0.9062 - loss: 0.3020 - roc_auc: 0.9922

 14/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9359 - loss: 0.2298 - roc_auc: 0.9813 

 26/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9338 - loss: 0.2202 - roc_auc: 0.9818

 39/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9351 - loss: 0.2143 - roc_auc: 0.9825

 51/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9369 - loss: 0.2104 - roc_auc: 0.9830

 65/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9382 - loss: 0.2086 - roc_auc: 0.9832

 79/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9388 - loss: 0.2083 - roc_auc: 0.9832

 93/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9389 - loss: 0.2091 - roc_auc: 0.9830

108/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9389 - loss: 0.2095 - roc_auc: 0.9829

116/116 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.9379 - loss: 0.2095 - roc_auc: 0.9827 - val_accuracy: 0.8546 - val_loss: 0.3549 - val_roc_auc: 0.9173 - learning_rate: 8.1000e-06


Epoch 39/100


  1/116 ━━━━━━━━━━━━━━━━━━━━ 2s 22ms/step - accuracy: 1.0000 - loss: 0.0598 - roc_auc: 1.0000

 15/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9501 - loss: 0.1598 - roc_auc: 0.9909 

 29/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9392 - loss: 0.1835 - roc_auc: 0.9870

 43/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9356 - loss: 0.1925 - roc_auc: 0.9856

 57/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9334 - loss: 0.1978 - roc_auc: 0.9848

 71/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9329 - loss: 0.2004 - roc_auc: 0.9845

 86/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9333 - loss: 0.2012 - roc_auc: 0.9845

100/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9336 - loss: 0.2015 - roc_auc: 0.9845

115/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9336 - loss: 0.2018 - roc_auc: 0.9845

116/116 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.9346 - loss: 0.2012 - roc_auc: 0.9852 - val_accuracy: 0.8382 - val_loss: 0.3606 - val_roc_auc: 0.9200 - learning_rate: 8.1000e-06


Epoch 40/100


  1/116 ━━━━━━━━━━━━━━━━━━━━ 3s 26ms/step - accuracy: 0.8750 - loss: 0.3133 - roc_auc: 0.9921

 14/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9277 - loss: 0.2165 - roc_auc: 0.9876 

 29/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9283 - loss: 0.2154 - roc_auc: 0.9856

 43/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9298 - loss: 0.2119 - roc_auc: 0.9854

 58/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9316 - loss: 0.2092 - roc_auc: 0.9853

 72/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9326 - loss: 0.2083 - roc_auc: 0.9851

 86/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9333 - loss: 0.2085 - roc_auc: 0.9848

100/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9336 - loss: 0.2091 - roc_auc: 0.9844

114/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9338 - loss: 0.2095 - roc_auc: 0.9842

116/116 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.9371 - loss: 0.2091 - roc_auc: 0.9828 - val_accuracy: 0.8546 - val_loss: 0.3520 - val_roc_auc: 0.9175 - learning_rate: 8.1000e-06


Epoch 41/100


  1/116 ━━━━━━━━━━━━━━━━━━━━ 2s 23ms/step - accuracy: 0.9688 - loss: 0.1341 - roc_auc: 1.0000

 15/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9509 - loss: 0.1690 - roc_auc: 0.9932 

 29/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9477 - loss: 0.1851 - roc_auc: 0.9896

 43/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9446 - loss: 0.1932 - roc_auc: 0.9877

 56/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9422 - loss: 0.1978 - roc_auc: 0.9866

 70/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9407 - loss: 0.2005 - roc_auc: 0.9859

 85/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9404 - loss: 0.2017 - roc_auc: 0.9855

100/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9404 - loss: 0.2022 - roc_auc: 0.9853

114/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9405 - loss: 0.2024 - roc_auc: 0.9852

116/116 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.9409 - loss: 0.2029 - roc_auc: 0.9843 - val_accuracy: 0.8508 - val_loss: 0.3541 - val_roc_auc: 0.9188 - learning_rate: 2.4300e-06


Epoch 42/100


  1/116 ━━━━━━━━━━━━━━━━━━━━ 2s 23ms/step - accuracy: 0.9375 - loss: 0.2132 - roc_auc: 0.9870

 16/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9411 - loss: 0.2130 - roc_auc: 0.9861 

 30/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9342 - loss: 0.2229 - roc_auc: 0.9827

 45/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9318 - loss: 0.2254 - roc_auc: 0.9816

 60/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9313 - loss: 0.2248 - roc_auc: 0.9814

 75/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9313 - loss: 0.2235 - roc_auc: 0.9814

 89/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9317 - loss: 0.2224 - roc_auc: 0.9815

103/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9321 - loss: 0.2213 - roc_auc: 0.9815

116/116 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.9325 - loss: 0.2151 - roc_auc: 0.9818 - val_accuracy: 0.8407 - val_loss: 0.3581 - val_roc_auc: 0.9195 - learning_rate: 2.4300e-06


Epoch 43/100


  1/116 ━━━━━━━━━━━━━━━━━━━━ 2s 23ms/step - accuracy: 0.8750 - loss: 0.4002 - roc_auc: 0.9372

 15/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9401 - loss: 0.2295 - roc_auc: 0.9805 

 29/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9440 - loss: 0.2142 - roc_auc: 0.9839

 43/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9414 - loss: 0.2146 - roc_auc: 0.9838

 58/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9392 - loss: 0.2158 - roc_auc: 0.9835

 72/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9384 - loss: 0.2160 - roc_auc: 0.9834

 86/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9383 - loss: 0.2149 - roc_auc: 0.9835

 99/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9381 - loss: 0.2139 - roc_auc: 0.9835

112/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9379 - loss: 0.2132 - roc_auc: 0.9836

116/116 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.9365 - loss: 0.2059 - roc_auc: 0.9842 - val_accuracy: 0.8445 - val_loss: 0.3568 - val_roc_auc: 0.9188 - learning_rate: 2.4300e-06


Epoch 44/100


  1/116 ━━━━━━━━━━━━━━━━━━━━ 2s 23ms/step - accuracy: 0.9688 - loss: 0.1150 - roc_auc: 1.0000

 16/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9515 - loss: 0.1713 - roc_auc: 0.9893 

 31/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9398 - loss: 0.1945 - roc_auc: 0.9857

 46/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9369 - loss: 0.2008 - roc_auc: 0.9847

 60/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9355 - loss: 0.2028 - roc_auc: 0.9843

 75/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9350 - loss: 0.2033 - roc_auc: 0.9842

 89/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9348 - loss: 0.2032 - roc_auc: 0.9842

103/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9347 - loss: 0.2037 - roc_auc: 0.9841

116/116 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.9352 - loss: 0.2065 - roc_auc: 0.9834 - val_accuracy: 0.8432 - val_loss: 0.3572 - val_roc_auc: 0.9194 - learning_rate: 2.4300e-06


Epoch 45/100


  1/116 ━━━━━━━━━━━━━━━━━━━━ 2s 23ms/step - accuracy: 0.8438 - loss: 0.2495 - roc_auc: 0.9615

 16/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9123 - loss: 0.2257 - roc_auc: 0.9770 

 30/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9245 - loss: 0.2141 - roc_auc: 0.9804

 45/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9285 - loss: 0.2104 - roc_auc: 0.9817

 59/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9304 - loss: 0.2091 - roc_auc: 0.9822

 69/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9313 - loss: 0.2083 - roc_auc: 0.9824

 80/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9317 - loss: 0.2081 - roc_auc: 0.9825

 93/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9320 - loss: 0.2083 - roc_auc: 0.9826

107/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9321 - loss: 0.2086 - roc_auc: 0.9826

116/116 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.9357 - loss: 0.2057 - roc_auc: 0.9837 - val_accuracy: 0.8407 - val_loss: 0.3581 - val_roc_auc: 0.9191 - learning_rate: 2.4300e-06


Epoch 46/100


  1/116 ━━━━━━━━━━━━━━━━━━━━ 2s 25ms/step - accuracy: 0.9688 - loss: 0.1341 - roc_auc: 0.9952

 14/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9530 - loss: 0.1872 - roc_auc: 0.9887 

 27/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9464 - loss: 0.2006 - roc_auc: 0.9859

 40/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9433 - loss: 0.2046 - roc_auc: 0.9849

 53/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9415 - loss: 0.2065 - roc_auc: 0.9843

 65/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9413 - loss: 0.2059 - roc_auc: 0.9843

 78/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9414 - loss: 0.2054 - roc_auc: 0.9842

 91/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9418 - loss: 0.2045 - roc_auc: 0.9843

104/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9420 - loss: 0.2037 - roc_auc: 0.9843

116/116 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.9419 - loss: 0.2014 - roc_auc: 0.9840 - val_accuracy: 0.8432 - val_loss: 0.3571 - val_roc_auc: 0.9192 - learning_rate: 1.0000e-06


Epoch 47/100


  1/116 ━━━━━━━━━━━━━━━━━━━━ 2s 25ms/step - accuracy: 0.8125 - loss: 0.4574 - roc_auc: 0.8874

 13/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8723 - loss: 0.3284 - roc_auc: 0.9494 

 25/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.8988 - loss: 0.2806 - roc_auc: 0.9643

 38/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9104 - loss: 0.2596 - roc_auc: 0.9706

 51/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9161 - loss: 0.2503 - roc_auc: 0.9733

 64/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9198 - loss: 0.2441 - roc_auc: 0.9750

 78/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9230 - loss: 0.2388 - roc_auc: 0.9762

 91/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9255 - loss: 0.2349 - roc_auc: 0.9771

105/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9274 - loss: 0.2318 - roc_auc: 0.9778

116/116 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.9382 - loss: 0.2154 - roc_auc: 0.9813 - val_accuracy: 0.8420 - val_loss: 0.3576 - val_roc_auc: 0.9192 - learning_rate: 1.0000e-06


Epoch 48/100


  1/116 ━━━━━━━━━━━━━━━━━━━━ 2s 24ms/step - accuracy: 0.9062 - loss: 0.2789 - roc_auc: 0.9773

 16/116 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.9496 - loss: 0.2044 - roc_auc: 0.9854 

 30/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9485 - loss: 0.1972 - roc_auc: 0.9867

 44/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9482 - loss: 0.1947 - roc_auc: 0.9869

 58/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9481 - loss: 0.1937 - roc_auc: 0.9868

 72/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9484 - loss: 0.1925 - roc_auc: 0.9868

 86/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9487 - loss: 0.1914 - roc_auc: 0.9868

100/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9490 - loss: 0.1906 - roc_auc: 0.9869

115/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9491 - loss: 0.1902 - roc_auc: 0.9869

116/116 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.9506 - loss: 0.1881 - roc_auc: 0.9868 - val_accuracy: 0.8483 - val_loss: 0.3557 - val_roc_auc: 0.9188 - learning_rate: 1.0000e-06


Epoch 49/100


  1/116 ━━━━━━━━━━━━━━━━━━━━ 2s 23ms/step - accuracy: 0.9062 - loss: 0.2376 - roc_auc: 0.9662

 16/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9322 - loss: 0.1821 - roc_auc: 0.9869 

 31/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9343 - loss: 0.1872 - roc_auc: 0.9862

 45/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9334 - loss: 0.1929 - roc_auc: 0.9853

 59/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9330 - loss: 0.1955 - roc_auc: 0.9850

 73/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9324 - loss: 0.1977 - roc_auc: 0.9847

 87/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9328 - loss: 0.1983 - roc_auc: 0.9846

101/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9331 - loss: 0.1987 - roc_auc: 0.9845

115/116 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.9335 - loss: 0.1991 - roc_auc: 0.9845

116/116 ━━━━━━━━━━━━━━━━━━━━ 1s 5ms/step - accuracy: 0.9379 - loss: 0.2009 - roc_auc: 0.9841 - val_accuracy: 0.8483 - val_loss: 0.3561 - val_roc_auc: 0.9188 - learning_rate: 1.0000e-06


training_curves.png guardado


## Celda 12 — Evaluación en test set

El test set (15% del total, ~794 muestras) no fue visto por el modelo
en ningún momento — ni para entrenar ni para ajustar hiperparámetros.

| Métrica | Objetivo | Razón |
|---------|----------|-------|
| Recall sismo | ≥ 0.85 | Un falso negativo (sismo no detectado) es inaceptable |
| F1 sismo | ≥ 0.80 | Balance precision/recall |
| Accuracy global | > 0.80 | Referencia; el recall sismo tiene prioridad |

Si el recall sismo cae por debajo de 0.80, considerar:
- Bajar el umbral de clasificación de 0.5 a 0.3 (tradeoff: más falsos positivos)
- Aumentar `class_weight[1]` y volver a entrenar
- Agregar PGA como feature adicional (Fase 2b)


In [7]:
y_prob = model.predict(X_test_n, verbose=0).ravel()
y_pred = (y_prob > 0.5).astype(int)

print("=" * 55)
print(classification_report(y_test, y_pred,
                             target_names=['0_Ruido', '1_Sismo'], digits=4))
print("=" * 55)

disp = ConfusionMatrixDisplay(confusion_matrix(y_test, y_pred),
                               display_labels=['Ruido', 'Sismo'])
disp.plot(cmap='Blues')
plt.title('Confusion Matrix — Test Set (umbral = 0.5)')
plt.savefig(MODEL_DIR / 'confusion_matrix.png', dpi=100)
plt.close()
print("confusion_matrix.png guardado")


              precision    recall  f1-score   support

     0_Ruido     0.8843    0.8682    0.8762       493
     1_Sismo     0.7903    0.8140    0.8020       301

    accuracy                         0.8476       794
   macro avg     0.8373    0.8411    0.8391       794
weighted avg     0.8487    0.8476    0.8480       794

confusion_matrix.png guardado


## Celda 14 — Exportación a TFLite INT8

**¿Por qué INT8?**
- Los pesos FP32 (4 bytes/valor) se comprimen a INT8 (1 byte) → ~4× menos espacio.
- La inferencia en microcontroladores sin FPU dedicada es más rápida con enteros.
- El ESP32 (Xtensa LX6) no tiene unidad SIMD vectorial, pero TFLite Micro
  usa kernels de multiplicación optimizados para INT8.

**Configuración elegida: pesos INT8, I/O float32**

Se cuantizan los pesos y activaciones internas a INT8, pero la entrada y
salida del modelo permanecen en `float32`. Esto simplifica el firmware:
el ESP32 pasa directamente el array normalizado de floats sin necesitar
calcular manualmente los parámetros de cuantización de entrada
(`scale` y `zero_point`) que varían por capa.

**`representative_dataset`:** TFLite necesita 200–500 muestras reales del
training set para calibrar los rangos de activación de cada capa y asignar
los valores INT8 óptimos. Sin este paso, la cuantización sería estática
y perdería precisión.

| Archivo | Descripción |
|---------|-------------|
| `seismic_fp32.tflite` | Línea base sin cuantizar (referencia) |
| `seismic_int8.tflite` | Modelo listo para ESP32 (< 300 KB esperado) |


In [8]:
# ── FP32 baseline ─────────────────────────────────────────────────────
converter_fp32 = tf.lite.TFLiteConverter.from_keras_model(model)
tflite_fp32    = converter_fp32.convert()
(MODEL_DIR / 'seismic_fp32.tflite').write_bytes(tflite_fp32)
print(f"FP32 baseline : {len(tflite_fp32)/1024:.1f} KB")

# ── INT8 (pesos + activaciones), I/O permanece float32 ─────────────────
def repr_gen():
    """Muestras representativas para calibrar rangos de activación INT8."""
    for x in X_train_n[:300]:
        yield [x[np.newaxis].astype(np.float32)]

converter_int8 = tf.lite.TFLiteConverter.from_keras_model(model)
converter_int8.optimizations             = [tf.lite.Optimize.DEFAULT]
converter_int8.representative_dataset   = repr_gen
converter_int8.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS_INT8]
converter_int8.inference_input_type     = tf.float32
converter_int8.inference_output_type    = tf.float32

tflite_int8 = converter_int8.convert()
(MODEL_DIR / 'seismic_int8.tflite').write_bytes(tflite_int8)
print(f"INT8 cuantizado: {len(tflite_int8)/1024:.1f} KB")
print(f"Compresión     : {len(tflite_fp32)/len(tflite_int8):.1f}×")

limite_kb = 300
ok = '✓' if len(tflite_int8)/1024 < limite_kb else '✗ SUPERA LÍMITE'
print(f"Objetivo < {limite_kb} KB : {ok}")


INFO:tensorflow:Assets written to: /tmp/tmpeg_4ijae/assets


INFO:tensorflow:Assets written to: /tmp/tmpeg_4ijae/assets


Saved artifact at '/tmp/tmpeg_4ijae'. The following endpoints are available:

* Endpoint 'serve'
  args_0 (POSITIONAL_ONLY): TensorSpec(shape=(None, 65, 11, 3), dtype=tf.float32, name='keras_tensor')
Output Type:
  TensorSpec(shape=(None, 1), dtype=tf.float32, name=None)
Captures:
  129264525101008: TensorSpec(shape=(), dtype=tf.resource, name=None)
  129264525104272: TensorSpec(shape=(), dtype=tf.resource, name=None)
  129264525102928: TensorSpec(shape=(), dtype=tf.resource, name=None)
  129264525101392: TensorSpec(shape=(), dtype=tf.resource, name=None)
  129264525101968: TensorSpec(shape=(), dtype=tf.resource, name=None)
  129264525103888: TensorSpec(shape=(), dtype=tf.resource, name=None)
  129264525101776: TensorSpec(shape=(), dtype=tf.resource, name=None)
  129264525103120: TensorSpec(shape=(), dtype=tf.resource, name=None)
  129264525105424: TensorSpec(shape=(), dtype=tf.resource, name=None)
  129264525105616: TensorSpec(shape=(), dtype=tf.resource, name=None)
  129264525104656:

W0000 00:00:1778679815.479616  202028 tf_tfl_flatbuffer_helpers.cc:364] Ignored output_format.
W0000 00:00:1778679815.479634  202028 tf_tfl_flatbuffer_helpers.cc:367] Ignored drop_control_dependency.


I0000 00:00:1778679815.480462  202028 reader.cc:83] Reading SavedModel from: /tmp/tmpeg_4ijae
I0000 00:00:1778679815.481689  202028 reader.cc:52] Reading meta graph with tags { serve }
I0000 00:00:1778679815.481703  202028 reader.cc:147] Reading SavedModel debug info (if present) from: /tmp/tmpeg_4ijae
I0000 00:00:1778679815.491564  202028 mlir_graph_optimization_pass.cc:437] MLIR V1 optimization pass is not enabled
I0000 00:00:1778679815.493235  202028 loader.cc:236] Restoring SavedModel bundle.
I0000 00:00:1778679815.542936  202028 loader.cc:220] Running initialization op on SavedModel bundle at path: /tmp/tmpeg_4ijae
I0000 00:00:1778679815.560084  202028 loader.cc:471] SavedModel load for tags { serve }; Status: success: OK. Took 79632 microseconds.
I0000 00:00:1778679815.627261  202028 dump_mlir_util.cc:269] disabling MLIR crash reproducer, set env var `MLIR_CRASH_REPRODUCER_DIRECTORY` to enable.


FP32 baseline : 114.8 KB


INFO:tensorflow:Assets written to: /tmp/tmpoy3jqi9e/assets


INFO:tensorflow:Assets written to: /tmp/tmpoy3jqi9e/assets


Saved artifact at '/tmp/tmpoy3jqi9e'. The following endpoints are available:

* Endpoint 'serve'
  args_0 (POSITIONAL_ONLY): TensorSpec(shape=(None, 65, 11, 3), dtype=tf.float32, name='keras_tensor')
Output Type:
  TensorSpec(shape=(None, 1), dtype=tf.float32, name=None)
Captures:
  129264525101008: TensorSpec(shape=(), dtype=tf.resource, name=None)
  129264525104272: TensorSpec(shape=(), dtype=tf.resource, name=None)
  129264525102928: TensorSpec(shape=(), dtype=tf.resource, name=None)
  129264525101392: TensorSpec(shape=(), dtype=tf.resource, name=None)
  129264525101968: TensorSpec(shape=(), dtype=tf.resource, name=None)
  129264525103888: TensorSpec(shape=(), dtype=tf.resource, name=None)
  129264525101776: TensorSpec(shape=(), dtype=tf.resource, name=None)
  129264525103120: TensorSpec(shape=(), dtype=tf.resource, name=None)
  129264525105424: TensorSpec(shape=(), dtype=tf.resource, name=None)
  129264525105616: TensorSpec(shape=(), dtype=tf.resource, name=None)
  129264525104656:

/home/pinpa/Documents/Universidad/Trabajo_de_Grado_Seismology/PANdeMaiz/pan_env/lib/python3.12/site-packages/tensorflow/lite/python/convert.py:846: UserWarning: Statistics for quantized inputs were expected, but not specified; continuing anyway.
  warnings.warn(


INT8 cuantizado: 40.1 KB
Compresión     : 2.9×
Objetivo < 300 KB : ✓


W0000 00:00:1778679816.485282  202028 tf_tfl_flatbuffer_helpers.cc:364] Ignored output_format.
W0000 00:00:1778679816.485305  202028 tf_tfl_flatbuffer_helpers.cc:367] Ignored drop_control_dependency.
I0000 00:00:1778679816.485481  202028 reader.cc:83] Reading SavedModel from: /tmp/tmpoy3jqi9e
I0000 00:00:1778679816.486441  202028 reader.cc:52] Reading meta graph with tags { serve }
I0000 00:00:1778679816.486449  202028 reader.cc:147] Reading SavedModel debug info (if present) from: /tmp/tmpoy3jqi9e
I0000 00:00:1778679816.497292  202028 loader.cc:236] Restoring SavedModel bundle.
I0000 00:00:1778679816.541926  202028 loader.cc:220] Running initialization op on SavedModel bundle at path: /tmp/tmpoy3jqi9e
I0000 00:00:1778679816.560007  202028 loader.cc:471] SavedModel load for tags { serve }; Status: success: OK. Took 74536 microseconds.
fully_quantize: 0, inference_type: 6, input_inference_type: FLOAT32, output_inference_type: FLOAT32
W0000 00:00:1778679816.778197  202028 flatbuffer_expo

## Celda 16 — Cabeceras C para el firmware ESP32

Esta celda produce dos archivos `.h` listos para copiar a
`platform_acel/acel_ubuntu/src/`:

**`model.h`**  
Convierte `seismic_int8.tflite` a un arreglo C de bytes (`uint8_t[]`).
El ESP32 lo carga en flash (PROGMEM) y TFLite Micro lo interpreta en
tiempo de ejecución sin necesidad de sistema de archivos.

**`norm_constants.h`**  
Exporta las constantes de normalización calculadas en la Celda 7:
- `SEISMIC_NORM_MEAN` / `SEISMIC_NORM_STD` — para el Z-score en C++
- `SEISMIC_HANN_POWER` — suma de cuadrados de la ventana Hanning de 128
  puntos (idéntica a la usada por scipy), necesaria para que el PSD
  calculado en el ESP32 coincida numéricamente con el del entrenamiento.

> ⚠️ Si se re-entrena el modelo (nuevas épocas, más datos, hiperparámetros
> distintos), se deben regenerar y copiar **ambas** cabeceras. Las constantes
> de normalización cambian si el dataset de entrenamiento cambia.


In [9]:
import scipy.signal as signal

# Constante de ventana Hanning — idéntica a scipy.signal.get_window('hann', 128)
win        = signal.get_window('hann', 128)
HANN_POWER = float(np.sum(win ** 2))
print(f"HANN_POWER = {HANN_POWER:.8f}  (Σhann² de 128 puntos)")

# ── model.h ──────────────────────────────────────────────────────────────
data   = (MODEL_DIR / 'seismic_int8.tflite').read_bytes()
c_body = ', '.join(f'0x{b:02x}' for b in data)
(MODEL_DIR / 'model.h').write_text(
    f'#pragma once\n'
    f'// Auto-generado por seismic_cnn_trainer.ipynb — NO editar a mano\n'
    f'// Arquitectura : 1D-CNN  |  Cuantizacion : INT8 (I/O float32)\n'
    f'// Tamano       : {len(data)} bytes = {len(data)/1024:.1f} KB\n'
    f'// Input shape  : (1, 65, 12, 3) float32\n'
    f'// Output       : float32 score en [0=ruido, 1=sismo]\n'
    f'#include <cstdint>\n\n'
    f'const uint8_t  seismic_model_data[] = {{{c_body}}};\n'
    f'const uint32_t seismic_model_len    = {len(data)}u;\n',
    encoding='utf-8'
)
print(f"model.h generado ({len(data)/1024:.1f} KB)")

# ── norm_constants.h ─────────────────────────────────────────────────────
(MODEL_DIR / 'norm_constants.h').write_text(
    f'#pragma once\n'
    f'// Auto-generado por seismic_cnn_trainer.ipynb — NO editar a mano\n'
    f'// Aplicar en ESP32 ANTES de pasar el espectrograma al modelo:\n'
    f'//   x_norm = (log10(PSD + 1e-12) - MEAN) / STD\n\n'
    f'// Z-score global (calculado sobre el training set)\n'
    f'constexpr float SEISMIC_NORM_MEAN  = {NORM_MEAN:.8f}f;\n'
    f'constexpr float SEISMIC_NORM_STD   = {NORM_STD:.8f}f;\n\n'
    f'// Ventana Hanning 128 puntos — scipy.signal.get_window("hann", 128)\n'
    f'// PSD = |FFT|^2 / (FS * HANN_POWER)   (one-sided, igual que scipy)\n'
    f'constexpr float SEISMIC_HANN_POWER = {HANN_POWER:.8f}f;\n'
    f'constexpr float SEISMIC_FS         = 200.0f;\n\n'
    f'// Umbral de clasificacion (score > THRESHOLD -> sismo)\n'
    f'constexpr float SEISMIC_THRESHOLD  = 0.50f;\n',
    encoding='utf-8'
)
print("norm_constants.h generado")
print(f"\n-> Copiar ambos archivos a: platform_acel/acel_ubuntu/src/")


HANN_POWER = 48.00000000  (Σhann² de 128 puntos)
model.h generado (40.1 KB)
norm_constants.h generado

-> Copiar ambos archivos a: platform_acel/acel_ubuntu/src/
